python/ray/data/datasource/iceberg_datasource.py

```diff
@@ -0,0 +1,187 @@
+        catalog = self._get_catalog()
+        table = catalog.load_table(self.table_identifier)
+
+        data_scan = table.scan(
+            row_filter=self._row_filter,
+            selected_fields=self._selected_fields,
+            snapshot_id=self._snapshot_id,
+            **self._scan_kwargs,
+        )
+
+        return table, data_scan
+
+    def estimate_inmemory_data_size(self) -> Optional[int]:
+        # Get the plan files in this query
+        _, data_scan = self._get_table_and_data_scan()
+        return sum(task.length for task in data_scan.plan_files())
+
+    @staticmethod
+    def _distribute_tasks_into_equal_chunks(
+        plan_files: Iterable["FileScanTask"], n_chunks: int
+    ) -> List[List["FileScanTask"]]:
+        """
+        Implement a greedy knapsack algorithm to distribute the files in the scan
+        across tasks, based on their file size, as evenly as possible
+        """
+        chunks = [list() for _ in range(n_chunks)]
+        chunk_sizes = {chunk: 0 for chunk in range(n_chunks)}
+
+        # From largest to smallest, add the plan files ot the smallest chunk one at a
+        # time
+        for plan_file in sorted(plan_files, key=lambda f: f.length, reverse=True):
+            smallest_chunk = min(chunk_sizes, key=chunk_sizes.get)
+            chunks[smallest_chunk].append(plan_file)
+            # Update the size of the chunk after adding the file to it
+            chunk_sizes[smallest_chunk] += plan_file.length
+
+        return chunks
+
+    def get_read_tasks(self, parallelism: int) -> List[ReadTask]:
+        from pyiceberg.io import pyarrow as pyi_pa_io
+
+        def _get_read_task(
+            tasks: Iterable["FileScanTask"],
+            table_identifier: str,
+            schema: "Schema",
+            catalog_type: str = "glue",
+        ) -> Iterable[block.Block]:
+            # Closure so we can pass this callable as an argument to ReadTask
+
+            # Both the catalog and tbl attributes cannot be pickled, which means they
+            # must be instantiated within this function (as opposed to being attributes
+            # of the IcebergDatasource class)
+            catalog = self._get_catalog()
+            tbl = catalog.load_table(table_identifier)
+
+            for task in tasks:
+                # Use the PyIceberg API to read only a single task (specifically, a
+                # FileScanTask) - note that this is not as simple as reading a single
+                # parquet file, as there might be delete files, etc. associated, so we
+                # must use the PyIceberg API for the projection.
+                yield pyi_pa_io.project_table(
+                    tasks=[task],
+                    table=tbl,
+                    row_filter=self._row_filter,
+                    projected_schema=schema,
+                    case_sensitive=self._scan_kwargs.get("case_sensitive", True),
+                    limit=self._scan_kwargs.get("limit"),
+                )
```

Comment:
It would be great to eventually push the deletes and projection to the engine itself. Would it be possible to pass in the group of tasks?

```suggestion
            # Use the PyIceberg API to read only a single task (specifically, a
            # FileScanTask) - note that this is not as simple as reading a single
            # parquet file, as there might be delete files, etc. associated, so we
            # must use the PyIceberg API for the projection.
            yield pyi_pa_io.project_table(
                tasks=tasks,
                table=tbl,
                row_filter=self._row_filter,
                projected_schema=schema,
                case_sensitive=self._scan_kwargs.get("case_sensitive", True),
                limit=self._scan_kwargs.get("limit"),
            )
```

This will be much more efficient

Reply:
I see what you mean here, but I was mostly worried about (1) reducing parallelism too much and (2) pushing too much computation down to the pyiceberg engine, i.e., what if all of `[tasks]` is too big to be held within a single read task?